### ISSUE TITLE
Username/Password auth initialization 

### ISSUE THREAD

**Issue by bvlaar** [ejholmes/restforce#171]
I have a rails 4 application that is currently using Restforce to login to a Salesforce Auth via username and password authentication (we're using one salesforce account to access the org). 

When I look at the app's performance, I see the salesforce requests (mostly queries) taking the most time. However, the request is broken down into two components, a login component and the actual request. The login component seems to be 50% of the total request time. 

The login is a post to: 
Net::HTTP[login.salesforce.com]: POST	
The response
Net::HTTP[na10.salesforce.com]: GET

I'm calling the following in every controller action:
client = Restforce.new
client.query(query string here)

Is there a way for me to initialize the app with client = Restforce.new in a config file and then import 'Client' into my controllers to prevent the app from logging in on every controller action that involves a Salesforce query? Would I be better off using oAuth?

**Comment by theSteveMitchell** [comment]
TL;DR try caching the auth_token and pass it to Restforce, and then update the cache after the query finishes:
```ruby 
def query(str)
    @client ||= Restforce.new(:oauth_token => Rails.cache.read('salesforceforce_oauth_token'))
    result = @client.query(str)
    Rails.cache.write('salesforceforce_oauth_token', client.options[:oauth_token])
    result
end
```
I had a similar issue with an app I wrote that also used a single application account, and forwarded a query to SF on almost every request.  The authentication phase of the request tok 80-90% of the total query time, and was incurred every time I Restforce.new'd

I found out that SF auth actually incurred 4-6 round trip requests to SF (a few requests to find the right instance server url, then a few requests for handshake).  The end result of the 6 roundtrips was the auth token.  I tried caching the auth token (we use Memcached, but any cache will do).  When you pass the token to Restforce, it will pass the existing token on the query request.  And as long as the token isn't expired, you'll get your query results with a single GET request to Salesforce

If the cached auth token has expired, then Restforce will automatically attempt to re-authenticate with your username and password, and will fetch a new auth token.  This is why we update the cache after a successful query.  

To be honest, I'm not sure if caching the entire Restforce.client instance works or not.  I think that I tried that, but found that it was still re-authenticating every time.  Would be cool if you could try it and report back.

With this approach, the average response time between my app and salesforce went from about 1200ms to about 120ms.

**Comment by bvlaar** [comment]
@theSteveMitchell Caching is a good idea. I need to try this.

One thing I want to clarify is that I am using the Username / Password auth method of Restforce (Restforce seems to differentiate oAuth and the username/password methods in the docs). I can easily switch to oAuth to try your method. My Restforce config passes the following into each request to auth:

    client = Restforce.new :username => 'foo',
      :password       => 'bar',
      :security_token => 'security token',
      :client_id      => 'client_id',
      :client_secret  => 'client_secret'

Appreciate the response, I'll report back with some findings shortly.

**Comment by theSteveMitchell** [comment]
Ah, you're right.  I simply meant to point out was that even with username/password authentication, the handshake will give you an oauth_token, and should still work.  I'm using the exact same auth method you are, so it should be good.

**Comment by myorukoglu** [comment]
@theSteveMitchell Thank you for the insight!

**Comment by theSteveMitchell** [comment]
@bvlaar Can we close this?

**Comment by bvlaar** [comment]
Yes, this was a big help. Thank you!

**Comment by antwonlee** [comment]
That's a significant enhancement in performance.  I am getting around 1000ms as well, and trying to look for cache solution.  I am going to try this and see if my app improves.  Thanks for the tip @theSteveMitchell

**Comment by theSteveMitchell** [comment]
@bvlaar can you close this?
